crypto/crypto.go

```diff
@@ -0,0 +1,153 @@
+func New(impl consensus.CryptoBase) consensus.Crypto {
+	return &crypto{CryptoBase: impl}
+}
+
+// InitConsensusModule gives the module a reference to the Modules object.
+// It also allows the module to set module options using the OptionsBuilder.
+func (c *crypto) InitConsensusModule(mods *consensus.Modules, cfg *consensus.OptionsBuilder) {
+	c.mods = mods
+	if mod, ok := c.CryptoBase.(consensus.Module); ok {
+		mod.InitConsensusModule(mods, cfg)
+	}
+}
+
+// CreatePartialCert signs a single block and returns the partial certificate.
+func (c crypto) CreatePartialCert(block *consensus.Block) (cert consensus.PartialCert, err error) {
+	sig, err := c.Sign(block.ToBytes())
+	if err != nil {
+		return consensus.PartialCert{}, err
+	}
+	return consensus.NewPartialCert(sig, block.Hash()), nil
+}
+
+// CreateQuorumCert creates a quorum certificate from a list of partial certificates.
+func (c crypto) CreateQuorumCert(block *consensus.Block, signatures []consensus.PartialCert) (cert consensus.QuorumCert, err error) {
+	// genesis QC is always valid.
+	if block.Hash() == consensus.GetGenesis().Hash() {
+		return consensus.NewQuorumCert(nil, 0, consensus.GetGenesis().Hash()), nil
+	}
+	sigs := make([]consensus.QuorumSignature, 0, len(signatures))
+	for _, sig := range signatures {
+		sigs = append(sigs, sig.Signature())
+	}
+	sig, err := c.Combine(sigs...)
+	if err != nil {
+		return consensus.QuorumCert{}, err
+	}
+	return consensus.NewQuorumCert(sig, block.View(), block.Hash()), nil
+}
+
+// CreateTimeoutCert creates a timeout certificate from a list of timeout messages.
+func (c crypto) CreateTimeoutCert(view consensus.View, timeouts []consensus.TimeoutMsg) (cert consensus.TimeoutCert, err error) {
+	// view 0 is always valid.
+	if view == 0 {
+		return consensus.NewTimeoutCert(nil, 0), nil
+	}
+	sigs := make([]consensus.QuorumSignature, 0, len(timeouts))
+	for _, timeout := range timeouts {
+		sigs = append(sigs, timeout.ViewSignature)
+	}
+	sig, err := c.Combine(sigs...)
+	if err != nil {
+		return consensus.TimeoutCert{}, err
+	}
+	return consensus.NewTimeoutCert(sig, view), nil
+}
+
+// CreateAggregateQC creates an AggregateQC from the given timeout messages.
+func (c crypto) CreateAggregateQC(view consensus.View, timeouts []consensus.TimeoutMsg) (aggQC consensus.AggregateQC, err error) {
+	qcs := make(map[hotstuff.ID]consensus.QuorumCert)
+	sigs := make([]consensus.QuorumSignature, 0, len(timeouts))
+	for _, timeout := range timeouts {
+		if qc, ok := timeout.SyncInfo.QC(); ok {
+			qcs[timeout.ID] = qc
+		}
+		if timeout.MsgSignature != nil {
+			sigs = append(sigs, timeout.MsgSignature)
+		}
+	}
+	sig, err := c.Combine(sigs...)
+	if err != nil {
+		return consensus.AggregateQC{}, err
+	}
+	return consensus.NewAggregateQC(qcs, sig, view), nil
+}
+
+// VerifyPartialCert verifies a single partial certificate.
+func (c crypto) VerifyPartialCert(cert consensus.PartialCert) bool {
+	block, ok := c.mods.BlockChain().Get(cert.BlockHash())
```

Comment:
Wouldn't it be better to pass in `block` to the function, instead of accessing the blockchain in the crypto package, causing an implicit dependency?  Same applies to Line 109. Maybe this is fixed in #61?

Reply:
I guess it is a responsibility problem. Should the caller of `VerifyPartialCert` be responsible for finding the block, or should the crypto implementation itself do it? I prefer it the way it is because it makes the crypto interface easier to use.